[TG-1058] Fixes parsing errors relating inner classes on generic classes

thk123 opened the pull request:
When parsing types that were inner classes inside generic classes

```java
class Outer<T> {
  class Inner { }
  void foo(Inner input) {} 
  Inner field;
}
```
Then you end up with singatures like:

```
LOuter<TT;>Inner;
```

The parsing was only parsing up the the first `;` or `<` meaning we missed out on the inner class. 

This PR resolves this (and a linked issue of if the inner class is generic parsing the generic details from within it). Most of this PR is in the unit tests, so I suggest reviewing commit by commit. I will first squash the commits a bit. Then re-run clang-format to tidy them up.

Commit:
Adding utilities for checking types in unit tests

Commit:
Unit test reproducing the bug described in TG-1058

When parsing fields that are of the inner class type, but the outer
class is generic, we wrongly parse this a pointer to the outer type.
This test demonstrates this problem.

Commit:
Added function for getting the full class name of a type

Skip over the signature when building the class name.

Commit:
Parse the whole class when generics are present

Previously when a generic signature was parsed, inner classes were
ignored. This resolves this by first building up a complete class name.
This also resolves the fact that inner classes in signatures are denoted
by . rather than $ and so replaces the . with $ to be consistent.

Commit:
Refactored out the code for erasing generic types

Commit:
Don't crash when found an invalid reference type

This is consistent with the original behaviour which would return a
nil_typet

Commit:
Adding utilities for code_type structures

Commit:
Adding tests for parsing methods of generic classes with inner classes

Commit:
Hide methods relating to getting multiple type variables for specific param

It is not possible for a generic parameter to either have multiple names
or have multiple instantiated types so the public interface should only
return one value.

Commit:
Add method for finding the matching ; corresponding to a reference L

Commit:
Correctly parse multiple input parameters for functions

When there is a function on a generic type whose params are inner
classes we would wrongly parse only up to the inner class dot for the
name. This resulted in wrong parameters causing us to fall back on the
descriptor unnecessarily.

Instead we should find the semi-colon that matches the L for each
parameter before moving on to the next parameter.

Commit:
Pulled out method for parsing a list of multiple types

Commit:
Parse multiple nested generic type informations

Use the logic to parse multiple parameters for generic type parameters.

Commit:
Extended tests to verify the generic information is being parsed correctly.

Commit:
Simplifying method names for unit test

Commit:
Adding doubly nested generic tests

thk123 commented:
This is based on an old version of a develop (49554177f8f467ea0bbbf9185ab68739694e8013) which is the current version on TG. This should be bumped before this is merged.

smowton commented on src/java_bytecode/java_types.cpp:
```diff
@@ -167,6 +167,218 @@ exprt java_bytecode_promotion(const exprt &expr)
+
+  // parse contained types, can be either type variables, starting with T
+  // or instantiated types
+  const std::vector<typet> &params =
```
Don't take by ref (it happens to work thanks to automagic lifetime extension, but it suggests something different to what's actually happening)

thk123 replied:
I though move semantics meant this was OK, but updated

smowton commented on src/java_bytecode/java_types.cpp:
```diff
@@ -167,6 +167,218 @@ exprt java_bytecode_promotion(const exprt &expr)
+
+  std::transform(params.begin(), params.end(),
+                 std::back_inserter(generic_type.generic_type_variables()),
+                 [](const typet &type) -> java_generic_parametert {
```
Kick this out into a non-member static function, it's too long for an inline lambda

thk123 replied:
My feeling is it is replacing the body of a for loop and therefore clearer to have it inline (particularly after applying below suggestions of returning directly). Can do if you disagree

smowton replied:
It's actually a lot better now that its indentation is fixed

smowton commented on src/java_bytecode/java_types.cpp:
```diff
@@ -167,6 +167,218 @@ exprt java_bytecode_promotion(const exprt &expr)
+                 std::back_inserter(generic_type.generic_type_variables()),
+                 [](const typet &type) -> java_generic_parametert {
+                   if (is_java_generic_parameter(type)) {
+                     const java_generic_parametert &gen_type =
```
Just return it directly?

smowton commented on src/java_bytecode/java_types.cpp:
```diff
@@ -167,6 +167,218 @@ exprt java_bytecode_promotion(const exprt &expr)
+                   } else {
+                     INVARIANT(is_reference(type),
+                               "All generic parameters should be references");
+                     java_generic_inst_parametert inst_type(
```
Similarly, might as well return directly

smowton commented on src/java_bytecode/java_types.cpp:
```diff
@@ -167,6 +167,218 @@ exprt java_bytecode_promotion(const exprt &expr)
+/// \return The input string with everything between angle brackets removed
+std::string erase_type_arguments(const std::string &src) {
+  std::string class_name = src;
+  std::size_t f_pos = class_name.find('<', 1);
```
To avoid duplcating this at line 234, turn this into `while(true)` and simply put the break condition in the middle of the loop

thk123 replied:
You mean like:
```C++
  std::size_t f_pos;
  while(true)
  {
    f_pos = class_name.find('<', 1);
    if(f_pos==std::string::npos) 
    {
      break;
    }
    std::size_t e_pos = find_closing_delimiter(class_name, f_pos, '<', '>');
    if(e_pos == std::string::npos)
      throw unsupported_java_class_signature_exceptiont("recursive generic");

    // erase generic information between brackets
    class_name.erase(f_pos, e_pos - f_pos + 1);
  }
```
I think this makes the code more confusing to follow is it now not as clear what the loop is looping over.

smowton commented on src/java_bytecode/java_types.cpp:
```diff
@@ -167,6 +167,218 @@ exprt java_bytecode_promotion(const exprt &expr)
+
+  class_name = erase_type_arguments(class_name);
+
+  std::replace(class_name.begin(), class_name.end(), '.', '$');
```
Suggest giving an example (in a comment) of a typical typename containing `.` and `/`, and how it appears afterwards

smowton commented on src/java_bytecode/java_types.cpp:
```diff
@@ -167,6 +167,218 @@ exprt java_bytecode_promotion(const exprt &expr)
+                                                  size_t starting_point = 0);
+
+/// Given a substring of a descriptor or signature that contains 1 or more types
+/// parse out the inidividual types. This is used for parsing the parameters of
```
Typo individual

smowton commented on src/java_bytecode/java_types.cpp:
```diff
@@ -167,6 +167,218 @@ exprt java_bytecode_promotion(const exprt &expr)
+
+/// Given a substring of a descriptor or signature that contains 1 or more types
+/// parse out the inidividual types. This is used for parsing the parameters of
+/// a function or the gneeric arguments contained within angle brackets.
```
typo generic

smowton commented on src/java_bytecode/java_types.cpp:
```diff
@@ -167,6 +167,218 @@ exprt java_bytecode_promotion(const exprt &expr)
+/// \param closing_bracket: For checking string is passed in as expected, the
+///   closing bracket (i.e. ')' or '>').
+/// \return A vector of types that the string represents
+std::vector<typet> parse_list_types(const std::string src,
```
Break line after `(`, indent params by two
e.g.
```
fghij(
  int x,
  int y)
```

thk123 replied:
Sorry messed up applying clang-format, this should now be auto-resolved.

smowton commented on src/java_bytecode/java_types.cpp:
```diff
@@ -167,6 +167,218 @@ exprt java_bytecode_promotion(const exprt &expr)
+                                    const std::string class_name_prefix,
+                                    const char opening_bracket,
+                                    const char closing_bracket) {
+  PRECONDITION(src[0] == opening_bracket);
```
Before these: precondition src.size()>=2

thk123 replied:
Also done for other functions with similar preconditions

smowton commented on src/java_bytecode/java_types.cpp:
```diff
@@ -167,6 +167,218 @@ exprt java_bytecode_promotion(const exprt &expr)
+    const typet &new_type = java_type_from_string(sub_str, class_name_prefix);
+    INVARIANT(new_type != nil_typet(), "Failed to parse type");
+
+    if (new_type.id() == ID_symbol) {
```
INVARIANT(new_type.id()!=ID_symbol)?

thk123 replied:
Sorry good spot - put in as erasing a condition I didn't think was relevant but forgot to change to an invariant.

thk123 replied:
In fact don't think it should be an invariant, but wanted to hard crash if anywhere that was true (it isn't particularly important here) so I've just removed it.

smowton commented on src/java_bytecode/java_types.cpp:
```diff
@@ -167,6 +167,218 @@ exprt java_bytecode_promotion(const exprt &expr)
+    size_t start = i;
+    while (i < src.size()) {
+      // parameter is an object type or instantiated generic type
+      if (src[i] == 'L') {
```
This looks very similar to `java_type_from_string`'s function parameter list parsing, can they be factored?

thk123 replied:
Unless I've misunderstood this comment, it is exactly that code and this is it being factored so can be used for parsing `<TT;Ljava/lang/Object;>` type strings.

smowton commented on src/java_bytecode/java_types.cpp:
```diff
@@ -167,6 +167,218 @@ exprt java_bytecode_promotion(const exprt &expr)
+///   generic types
+/// \return The reference type if parsed correctly, no value if parsing fails.
+optionalt<reference_typet>
+build_class_name(const std::string src, const std::string &class_name_prefix) {
```
Take input string by ref?

smowton commented on src/java_bytecode/java_types.cpp:
```diff
@@ -167,6 +167,218 @@ exprt java_bytecode_promotion(const exprt &expr)
+  symbol_typet symbol_type(identifier);
+  symbol_type.set(ID_C_base_name, container_class);
+
+  std::size_t f_pos = src.find('<', 1);
```
Same loop rotation recommendation as above -- prefer writing this update condition once and doing `while(true) { .... if(exit_condition) break; ... }` over duplicating it like this

thk123 replied:
See above discussion, will apply both changes if you want, otherwise neither

mgudemann commented on src/java_bytecode/java_types.cpp:
```diff
@@ -167,6 +167,243 @@ exprt java_bytecode_promotion(const exprt &expr)
     return typecast_exprt(expr, new_type);
 }
 
+std::vector<typet> parse_list_types(
```
shouldn't this be declared in the header file?

thk123 replied:
This is why I prefer classes even without state, as it allows declaring in the header without making a global function of everything that includes `java_types.h` I will move the declarations to the top of the file for now and raise a ticket to refactor this parsing into a class if this is satisfactory for you?

thk123 replied:
Done and raised as [[TG-1206](https://diffblue.atlassian.net/browse/TG-1206)]

mgudemann commented on src/java_bytecode/java_types.cpp:
```diff
@@ -167,6 +167,243 @@ exprt java_bytecode_promotion(const exprt &expr)
+  const std::vector<typet> &params =
+    parse_list_types(parameters, class_name_prefix, '<', '>');
+
+  CHECK_RETURN(params.size() > 0); // We should have at least one generic param
```
use `!..empty()`

smowton commented on src/java_bytecode/java_types.cpp:
```diff
@@ -167,6 +167,218 @@ exprt java_bytecode_promotion(const exprt &expr)
+            "There must be a semi-colon somewhere in the input");
+  size_t next_angle_bracket = src.find('<', starting_point);
+
+  while (next_angle_bracket < next_semi_colon) {
```
This might work, but I'd like to see some argument about why this exhaustively covers the type signature grammar... if this gets too tricky we should probably abandon manual parser writing and instead make a Bison or similar parse for the signature grammar (https://docs.oracle.com/javase/specs/jvms/se7/html/jvms-4.html#jvms-4.3.4)

This exists but is GPL'd: https://github.com/JetBrains/jdk8u_jdk/blob/master/src/share/classes/sun/reflect/generics/parser/SignatureParser.java

thk123 replied:
Yes I completely agree, unfortunately there isn't time at this juncture to do this. Here is why I'm convinced this is reasonable for now:

This is for parsing a reference type from either a signature or descriptor. If it is descriptor, then the grammar is exactly:
`L ClassName ;`

So since ClassName can't contain `;` or `<` we're done. 

(Which is why it worked until now). 

For signatures, it is more complicated:
```
L PackageSpecifier_opt SimpleClassTypeSignature ClassTypeSignatureSuffix* ;
```

The package specifier is just `identifier/identifier` so no chances of `;` or `<'`

Then the `SimpleClassTypeSignature` is 

```
Identifier TypeArguments_opt
```
Where type arguments must start with `<` (so this is the case where `next_angle_bracket < next_semi_colon` and must end with a corresponding `>` 

This is then optionally followed by a `.` indicating inner class and then repeat above argument. It is finally ended by a semi colon. 

I appreciate this is _not_ convincing, hopefully the extensive unit tests are *more* convincing - though I accept they are a long way from an exhaustive proof.

smowton reviewed (commented):
Looks tricky. Perhaps this is nasty enough that we need a real parser generator?

mgudemann commented on src/java_bytecode/java_types.cpp:
```diff
@@ -167,6 +167,243 @@ exprt java_bytecode_promotion(const exprt &expr)
+  {
+    std::size_t e_pos = find_closing_delimiter(class_name, f_pos, '<', '>');
+    if(e_pos == std::string::npos)
+      throw unsupported_java_class_signature_exceptiont("recursive generic");
```
I don't understand this, wouldn't this return `npos` if no corresponding closing `>` can be found, not if it is recursive generic?

thk123 replied:
You are right - however this case happens precisely when we have a recursive generic type. Given ```java_generic_type_from_string(<T:LSimpleRecursiveGeneric<TT;>;>Ljava/lang/Object;)``` calls erase type arguments with `T:LSimpleRecursiveGeneric<TT`  (https://diffblue.atlassian.net/browse/TG-610)

majakusber replied:
Even after recursive generic types are fixed, the exception here makes sense. How about changing it to something more general like `unsupported_java_class_signature_exceptiont("Failed to find generic signature closing delimiter");` or `unsupported_java_class_signature_exceptiont("Failed to find generic signature closing delimiter (or recursive generic)");` if we want to keep the reminder that recursives cause problem here

mgudemann commented on src/java_bytecode/java_types.cpp:
```diff
@@ -167,6 +167,243 @@ exprt java_bytecode_promotion(const exprt &expr)
+  return class_name;
+}
+
+size_t find_closing_semi_colon_for_reference_type(
```
why not in the header?

thk123 replied:
As above - let me know if you'd be happy with my proposed solution

mgudemann replied:
ok with me

mgudemann commented on src/java_bytecode/java_types.cpp:
```diff
@@ -167,6 +167,243 @@ exprt java_bytecode_promotion(const exprt &expr)
+  const std::string src,
+  size_t starting_point = 0);
+
+/// Given a substring of a descriptor or signature that contains 1 or more types
```
1->one

mgudemann commented on src/java_bytecode/java_types.cpp:
```diff
@@ -167,6 +167,243 @@ exprt java_bytecode_promotion(const exprt &expr)
+  std::vector<typet> type_list;
+  for(std::size_t i = 1; i < src.size() - 1; i++)
+  {
+    size_t start = i;
```
please document what the loop does

thk123 replied:
Added:
```
  // Loop over the types in the given string, parsing each one in turn
  // and adding to the type_list
```

mgudemann commented on src/java_bytecode/java_types.cpp:
```diff
@@ -167,6 +167,243 @@ exprt java_bytecode_promotion(const exprt &expr)
+    const typet &new_type = java_type_from_string(sub_str, class_name_prefix);
+    INVARIANT(new_type != nil_typet(), "Failed to parse type");
+
+    if(new_type.id() == ID_symbol)
```
why not an invariant with information string?

thk123 replied:
Removed as was only there as [previously we did](https://github.com/diffblue/cbmc/pull/1512/files#diff-ba71807a8e04db80ccb196e4eb6465feL281):

```
if(param.type().id()==ID_symbol)		
  param.type()=java_reference_type(param.type());
```

but afaict this never happens (wanted to verify this with a hard fail - but since the CI passes and I can't see a way for this to happen, I guess it is just legacy code)

mgudemann commented on src/java_bytecode/java_types.cpp:
```diff
@@ -167,6 +167,243 @@ exprt java_bytecode_promotion(const exprt &expr)
+    {
+      std::size_t e_pos = find_closing_delimiter(src, f_pos, '<', '>');
+      if(e_pos == std::string::npos)
+        throw unsupported_java_class_signature_exceptiont("recursive generic");
```
as above

thk123 replied:
We don't appear to have a test case that exercises this path so I will change the exception into a more prescriptive name

mgudemann commented on src/java_bytecode/java_types.cpp:
```diff
@@ -167,6 +167,243 @@ exprt java_bytecode_promotion(const exprt &expr)
+{
+  PRECONDITION(src[0] == 'L');
+  // ends on ;
+  if(src[src.size() - 1] != ';')
```
isn't this an invariant for any non-atomic class file level type ?

thk123 replied:
I thought so, changed this to an assertion and a test failed so I reverted back to original behaviour, but I can now no longer reproduce this, so perhaps I had something else wrong.

mgudemann commented on src/java_bytecode/java_types.cpp:
```diff
@@ -167,6 +167,243 @@ exprt java_bytecode_promotion(const exprt &expr)
+/// \return The string position corresponding to the matching ';'. For example:
+/// LA;, we would return 2. For LA<TT;>; we would return 7.
+/// See unit/java_bytecode/java_util_tests.cpp for more examples.
+size_t find_closing_semi_colon_for_reference_type(
```
would it make sense to adapt the `find_closing_delimiter` function to take a set of opening characters? Then using `find_closing_delimiters(src, end, "LT", `;`)` would return the `;` that corresponds to the first `L` and skip over any other `L`,`;` or `T`, `;` pair

thk123 replied:
I think the problem is you have two types of brackets `L`, `T`, which are closed by `;` and `<` which is closed by `>`. I suppose it would in fact work as `Lfoo<;>`; isn't legal, but presumably wouldn't bec caught if that did happen. Also `LL;` would fail (class named L)

thk123 replied:
And indeed the far more common `Lgeneric_class<TT;>;`

mgudemann replied:
the idea of `find_closing_delimiters` is to support hierarchic structures (maybe it's buggy?) and allow for free definition of closing and opening delimiter.
What I imagined was to adapt it in such a way that it would in this case
 - ignore `<` and `>`
 - increase the depth at `T` and `L`
 - decrease depth at `;`

agreed for `LL;`  or `LT;`, but with one lookahead or a Boolean that indicates that the last char was an `L` and "opened" the object name, we would be good, wouldn't we?
This parsing is slightly fragile and with one common function, we'd at least know where to correct an error and don't risk overseeing another place.

thk123 replied:
Surely if you ignore `<`, there is no difference between `LALFoo;` (a class called `ALFoo`) and `LA<LFoo;>;` and in the first case you won't find a matching `;`? 

I'll give it a go but I think this is a fudamentally different task (matching two isolated sets of brackets) and I think it's a bad idea to complicate `find_closing_delimiters`

mgudemann replied:
keep it like that then, with the overall goal to get in a real parser, I am ok with this

mgudemann commented on unit/java_bytecode/java_utils_test.cpp:
```diff
@@ -13,6 +13,7 @@
 
 #include <testing-utils/catch.hpp>
 
+#include <java_bytecode/java_types.cpp>
```
including the `cpp` file here? is this the reason for the definitions in the `cpp` file instead of the header ?

thk123 replied:
Vice-versa, including as declarations are not in the header, not in the header to avoid making the functions global. See earlier discussion and let me know if you're happy with my proposed solution (this will still include the cpp file)

Commit:
Assertion on unmatched ; for parsing reference types

Commit:
Tightened up exception for unknown handling

We don't have a test that exercises this exception so modified the
message to not mislead and say to do with recursive generic types.

Commit:
Fixed error in find_closing_delimiter

Assumed that the bracket was always an angle bracket

thk123 commented:
Thanks for the reviews - I've updated PR (didn't squash commits to make it easier to review changes - but indicated in commit message what they would be squashed into). 

Will address the liner issues tomorrow but let me know about the comments I've commented on @smowton @mgudemann

mgudemann commented:
@thk123 won't fight for the parsing stuff, but we should maybe think of using a parser generator and implement the grammar given in the spec.

majakusber commented on unit/java_bytecode/java_utils_test.cpp:
```diff
@@ -63,3 +64,160 @@ SCENARIO("Test that the generic signature delimiter lookup works reliably",
+  }
+  GIVEN("Signature: An generic inner class in a generic class")
+  {
+    std::string signature = "LClassName<TT;>.Inner<VV;>;";
```
Should it be TV instead of VV here? (similarly in multiple places in this scenario) I thought all type variables are given prefix T.

thk123 replied:
Good spot!

majakusber commented on src/java_bytecode/java_types.cpp:
```diff
@@ -196,6 +196,97 @@ std::string gather_full_class_name(const std::string &src)
+  {
+    std::size_t e_pos = find_closing_delimiter(src, f_pos, '<', '>');
+    if(e_pos == std::string::npos)
+      throw unsupported_java_class_signature_exceptiont("recursive generic");
```
Same as in the other file, the exception makes sense in general, not only for recursive generics, consider changing to something like ``unsupported_java_class_signature_exceptiont("Failed to find generic signature closing delimiter");``

thk123 commented:
@mgudemann @smowton I have raised a ticket [TG-1205](https://diffblue.atlassian.net/browse/TG-1205) to convert this conversion to using a parser generator since this logic is clearly confusing and error prone. 

I will try using `find_closing_delimiter` but other than that I think this is ready for re-review

Commit:
Added tests for bracket matcher to include different types of brackets

majakusber reviewed (approved):
Looks good, I only had three small things - one of them you already fixed (I was looking at it commit by commit).

thk123 commented:
Remaining linter errors are discrepancies between clang-format and `cpplint.py` - there is an on-going discussion as to what to do with these

smowton reviewed (approved):
LGTM

thk123 commented:
Rebased - just waiting on CI then good to merge
